chainservice/builder.go

```diff
@@ -0,0 +1,689 @@
+	"github.com/iotexproject/iotex-core/state/factory"
+	"github.com/iotexproject/iotex-election/committee"
+	"github.com/iotexproject/iotex-proto/golang/iotexrpc"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+// Builder is a builder to build chainservice
+type Builder struct {
+	cfg config.Config
+	cs  *ChainService
+}
+
+// NewBuilder creates a new chainservice builder
+func NewBuilder(cfg config.Config) *Builder {
+	builder := &Builder{cfg: cfg}
+	builder.createInstance()
+
+	return builder
+}
+
+// SetActionPool sets the action pool instance
+func (builder *Builder) SetActionPool(ap actpool.ActPool) *Builder {
+	builder.createInstance()
+	builder.cs.actpool = ap
+	return builder
+}
+
+// SetBlockchain sets the blockchain instance
+func (builder *Builder) SetBlockchain(bc blockchain.Blockchain) *Builder {
+	builder.createInstance()
+	builder.cs.chain = bc
+	return builder
+}
+
+// SetFactory sets the factory instance
+func (builder *Builder) SetFactory(f factory.Factory) *Builder {
+	builder.createInstance()
+	builder.cs.factory = f
+	return builder
+}
+
+// SetBlockDAO sets the blockdao instance
+func (builder *Builder) SetBlockDAO(bd blockdao.BlockDAO) *Builder {
+	builder.createInstance()
+	builder.cs.blockdao = bd
+	return builder
+}
+
+// SetP2PAgent sets the P2PAgent instance
+func (builder *Builder) SetP2PAgent(agent p2p.Agent) *Builder {
+	builder.createInstance()
+	builder.cs.p2pAgent = agent
+	return builder
+}
+
+// SetElectionCommittee sets the election committee instance
+func (builder *Builder) SetElectionCommittee(c committee.Committee) *Builder {
+	builder.createInstance()
+	builder.cs.electionCommittee = c
+	return builder
+}
+
+// SetBlockSync sets the block sync instance
+func (builder *Builder) SetBlockSync(bs blocksync.BlockSync) *Builder {
+	builder.createInstance()
+	builder.cs.blocksync = bs
+	return builder
+}
+
+// BuildForTest builds a chainservice for test purpose
+func (builder *Builder) BuildForTest() (*ChainService, error) {
+	builder.createInstance()
+	return builder.build(false, true)
+}
+
+// BuildForSubChain builds a chainservice for subchain
+func (builder *Builder) BuildForSubChain() (*ChainService, error) {
+	builder.createInstance()
+	return builder.build(true, false)
+}
+
+// Build builds a chainservice
+func (builder *Builder) Build() (*ChainService, error) {
+	builder.createInstance()
+	return builder.build(false, false)
+}
+
+func (builder *Builder) createInstance() {
+	if builder.cs == nil {
+		builder.cs = &ChainService{}
+	}
+}
+
+func (builder *Builder) buildFactory(forTest bool) error {
+	factory, err := builder.createFactory(forTest)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create state factory")
+	}
+	builder.cs.factory = factory
+	return nil
+}
+
+func (builder *Builder) createFactory(forTest bool) (factory.Factory, error) {
+	if builder.cs.factory != nil {
+		return builder.cs.factory, nil
+	}
+	if forTest {
```

Comment:
which factory should be returned when `forTest`  is true?

Reply:
The logic is not changed in this PR. You can check chainservice.New

Reply:
I mean L165 might never be run?